File: src/portfolio_manager/services/portfolio_service.py
"""Portfolio service for aggregating holdings data."""

from dataclasses import dataclass
from decimal import Decimal
from typing import TYPE_CHECKING

from portfolio_manager.models import Group, Stock
from portfolio_manager.repositories.group_repository import GroupRepository
from portfolio_manager.repositories.holding_repository import HoldingRepository
from portfolio_manager.repositories.stock_repository import StockRepository

if TYPE_CHECKING:
    from portfolio_manager.services.exchange.exchange_rate_service import (
        ExchangeRateService,
    )
    from portfolio_manager.services.price_service import PriceService


@dataclass
class StockHolding:
    """Stock with aggregated quantity."""

    stock: Stock
    quantity: Decimal


@dataclass
class StockHoldingWithPrice:
    """Stock with aggregated quantity and price information."""

    stock: Stock
    quantity: Decimal
    price: Decimal
    currency: str
    name: str
    value_krw: Decimal | None = None

    @property
    def value(self) -> Decimal:
        """Calculate total value (quantity × price)."""
        return self.quantity * self.price


@dataclass
class GroupHoldings:
    """Group with its stock holdings."""

    group: Group
    stock_holdings: list[StockHolding]


@dataclass
class PortfolioSummary:
    """Portfolio summary with price information."""

    holdings: list[tuple[Group, StockHoldingWithPrice]]
    total_value: Decimal


class PortfolioService:
    """Service for portfolio operations."""

    def __init__(
        self,
        group_repository: GroupRepository,
        stock_repository: StockRepository,
        holding_repository: HoldingRepository,
        price_service: "PriceService | None" = None,
        exchange_rate_service: "ExchangeRateService | None" = None,
    ):
        """Initialize service with repositories."""
        self.group_repository = group_repository
        self.stock_repository = stock_repository
        self.holding_repository = holding_repository
        self.price_service = price_service
        self.exchange_rate_service = exchange_rate_service

    def get_holdings_by_group(self) -> list[GroupHoldings]:
        """Get holdings aggregated by group."""
        groups = self.group_repository.list_all()
        aggregated_holdings = self.holding_repository.get_aggregated_holdings_by_stock()

        result = []
        for group in groups:
            stocks = self.stock_repository.list_by_group(group.id)
            stock_holdings = []
            for stock in stocks:
                quantity = aggregated_holdings.get(stock.id, Decimal("0"))
                if quantity > 0:
                    stock_holdings.append(StockHolding(stock=stock, quantity=quantity))
            result.append(GroupHoldings(group=group, stock_holdings=stock_holdings))

        return result

    def get_portfolio_summary(self) -> PortfolioSummary:
        """Get portfolio summary with valuations."""
        if self.price_service is None:
            raise ValueError("Price service is required for portfolio summary")

        groups = self.group_repository.list_all()
        aggregated_holdings = self.holding_repository.get_aggregated_holdings_by_stock()

        holdings = []
        total_value = Decimal("0")
        usd_krw_rate: Decimal | None = None

        for group in groups:
            stocks = self.stock_repository.list_by_group(group.id)
            for stock in stocks:
                quantity = aggregated_holdings.get(stock.id, Decimal("0"))
                if quantity > 0:
                    price, currency, name = self.price_service.get_stock_price(
                        stock.ticker
                    )
                    value_krw: Decimal | None = None
                    holding_value = quantity * price
                    if currency == "USD":
                        if self.exchange_rate_service is None:
                            raise ValueError(
                                "Exchange rate service is required for USD"
                            )
                        if usd_krw_rate is None:
                            usd_krw_rate = self.exchange_rate_service.get_usd_krw_rate()
                        value_krw = holding_value * usd_krw_rate
                    else:
                        value_krw = holding_value
                    holding_with_price = StockHoldingWithPrice(
                        stock=stock,
                        quantity=quantity,
                        price=price,
                        currency=currency,
                        name=name,
                        value_krw=value_krw,
                    )
                    holdings.append((group, holding_with_price))
                    if value_krw is not None:
                        total_value += value_krw

        return PortfolioSummary(holdings=holdings, total_value=total_value)

File: internal/services/portfolio_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"math"
	"sort"

	"github.com/kadragon/portfolio-manager/internal/datex"
	"github.com/kadragon/portfolio-manager/internal/ktime"
	"github.com/kadragon/portfolio-manager/internal/models"
	"github.com/kadragon/portfolio-manager/internal/numeric"
	"github.com/kadragon/portfolio-manager/internal/repositories"
	"github.com/kadragon/portfolio-manager/internal/uuidx"
)

// ErrNoPriceService is returned by GetPortfolioSummary when no PriceService is wired.
var ErrNoPriceService = errors.New("price service not configured")

// PortfolioService aggregates portfolio data across groups, stocks, and holdings.
type PortfolioService struct {
	groups       *repositories.GroupRepository
	stocks       *repositories.StockRepository
	holdings     *repositories.HoldingRepository
	accounts     *repositories.AccountRepository
	deposits     *repositories.DepositRepository
	priceService *PriceService
	exchangeRate *ExchangeRateService
}

// NewPortfolioService builds a PortfolioService.
// priceService and exchangeRate may be nil (only GetHoldingsByGroup available).
func NewPortfolioService(
	groups *repositories.GroupRepository,
	stocks *repositories.StockRepository,
	holdings *repositories.HoldingRepository,
	accounts *repositories.AccountRepository,
	deposits *repositories.DepositRepository,
	priceService *PriceService,
	exchangeRate *ExchangeRateService,
) *PortfolioService {
	return &PortfolioService{
		groups:       groups,
		stocks:       stocks,
		holdings:     holdings,
		accounts:     accounts,
		deposits:     deposits,
		priceService: priceService,
		exchangeRate: exchangeRate,
	}
}

// HasPriceService reports whether a PriceService is configured.
func (s *PortfolioService) HasPriceService() bool {
	return s.priceService != nil
}

// GetHoldingsByGroup returns all groups with their aggregated holdings (qty > 0).
// Groups with no holdings are included with an empty StockHoldings list.
func (s *PortfolioService) GetHoldingsByGroup(ctx context.Context) ([]models.GroupHoldings, error) {
	groups, err := s.groups.ListAll(ctx)
	if err != nil {
		return nil, err
	}

	allStocks, err := s.stocks.ListAll(ctx)
	if err != nil {
		return nil, err
	}

	aggregated, err := s.holdings.GetAggregatedByStock(ctx)
	if err != nil {
		return nil, err
	}

	stocksByGroup := make(map[uuidx.UUID][]models.Stock)
	for _, stock := range allStocks {
		stocksByGroup[stock.GroupID] = append(stocksByGroup[stock.GroupID], stock)
	}

	result := make([]models.GroupHoldings, 0, len(groups))
	for _, group := range groups {
		stocks := stocksByGroup[group.ID]
		stockHoldings := make([]models.StockHolding, 0)
		for _, stock := range stocks {
			qty, ok := aggregated[stock.ID]
			if !ok || qty.IsZero() {
				continue
			}
			stockHoldings = append(stockHoldings, models.StockHolding{
				Stock:    stock,
				Quantity: qty,
			})
		}
		result = append(result, models.GroupHoldings{
			Group:         group,
			StockHoldings: stockHoldings,
		})
	}
	return result, nil
}

// GetPortfolioSummary computes a full portfolio summary using DB-cached prices.
// Returns ErrNoPriceService if no PriceService is configured.
func (s *PortfolioService) GetPortfolioSummary(ctx context.Context) (*models.PortfolioSummary, error) {
	if s.priceService == nil {
		return nil, ErrNoPriceService
	}

	groups, err := s.groups.ListAll(ctx)
	if err != nil {
		return nil, err
	}

	allStocks, err := s.stocks.ListAll(ctx)
	if err != nil {
		return nil, err
	}

	aggregated, err := s.holdings.GetAggregatedByStock(ctx)
	if err != nil {
		return nil, err
	}

	stocksByGroup := make(map[uuidx.UUID][]models.Stock)
	for _, stock := range allStocks {
		stocksByGroup[stock.GroupID] = append(stocksByGroup[stock.GroupID], stock)
	}

	var usdKRW *numeric.Decimal
	pairs := make([]models.GroupHoldingPair, 0)
	totalStockValue := numeric.Zero

	for _, group := range groups {
		for _, stock := range stocksByGroup[group.ID] {
			qty, ok := aggregated[stock.ID]
			if !ok || qty.IsZero() {
				continue
			}
			preferredExchange := ""
			if stock.Exchange != nil {
				preferredExchange = *stock.Exchange
			}
			price, currency, name, exchange := s.priceService.GetStockPrice(ctx, stock.Ticker, preferredExchange)
			if name == "" {
				name = stock.Name
			}
			holdingValue := numeric.Wrap(qty.Mul(price.Decimal))

			var valueKRW *numeric.Decimal
			if currency == "USD" {
				if s.exchangeRate == nil {
					v := numeric.Zero
					valueKRW = &v
				} else {
					if usdKRW == nil {
						r := s.exchangeRate.GetUSDKRW()
						usdKRW = &r
					}
					v := numeric.Wrap(holdingValue.Mul(usdKRW.Decimal))
					valueKRW = &v
				}
			} else {
				v := holdingValue
				valueKRW = &v
			}

			totalStockValue = numeric.Wrap(totalStockValue.Add(valueKRW.Decimal))

			_ = exchange // stored in stock after Phase 8 sync
			pairs = append(pairs, models.GroupHoldingPair{
				Group: group,
				Holding: models.StockHoldingWithPrice{
					Stock:    stock,
					Quantity: qty,
					Price:    price,
					Currency: currency,
					Name:     name,
					ValueKRW: valueKRW,
				},
			})
		}
	}

	sort.Slice(pairs, func(i, j int) bool {
		vi, vj := numeric.Zero, numeric.Zero
		if pairs[i].Holding.ValueKRW != nil {
			vi = *pairs[i].Holding.ValueKRW
		}
		if pairs[j].Holding.ValueKRW != nil {
			vj = *pairs[j].Holding.ValueKRW
		}
		return vi.GreaterThan(vj.Decimal)
	})

	totalCash := numeric.Zero
	if s.accounts != nil {
		accs, err := s.accounts.ListAll(ctx)
		if err == nil {
			for _, a := range accs {
				totalCash = numeric.Wrap(totalCash.Add(a.CashBalance.Decimal))
			}
		}
	}

	totalInvested := numeric.Zero
	var firstDate *datex.Date
	if s.deposits != nil {
		deps, err := s.deposits.ListAll(ctx)
		if err == nil {
			for _, d := range deps {
				totalInvested = numeric.Wrap(totalInvested.Add(d.Amount.Decimal))
			}
		}
		firstDate, _ = s.deposits.GetFirstDepositDate(ctx)
	}

	totalAssets := numeric.Wrap(totalStockValue.Add(totalCash.Decimal))

	var returnRate *numeric.Decimal
	var annualizedReturn *numeric.Decimal
	if totalInvested.IsPositive() {
		r := numeric.Wrap(totalAssets.Sub(totalInvested.Decimal).Div(totalInvested.Decimal).Mul(hundred.Decimal))
		returnRate = &r

		if firstDate != nil {
			daysElapsed := int(ktime.Now().Sub(firstDate.Time).Hours() / 24)
			if daysElapsed > 0 {
				ratio, _ := totalAssets.Div(totalInvested.Decimal).Float64()
				annualizedRatio := math.Pow(ratio, 365.0/float64(daysElapsed))
				ar, _ := numeric.FromString(fmt.Sprintf("%.10f", (annualizedRatio-1)*100))
				annualizedReturn = &ar
			}
		}
	}

	return &models.PortfolioSummary{
		Holdings:             pairs,
		TotalValue:           totalStockValue,
		TotalStockValue:      totalStockValue,
		TotalCashBalance:     totalCash,
		TotalAssets:          totalAssets,
		TotalInvested:        totalInvested,
		ReturnRate:           returnRate,
		FirstDepositDate:     firstDate,
		AnnualizedReturnRate: annualizedReturn,
	}, nil
}

// ComputeGroupSummary aggregates holdings by group for the allocation table.
func ComputeGroupSummary(summary *models.PortfolioSummary) []models.GroupSummaryRow {
	totals := make(map[uuidx.UUID]numeric.Decimal)
	groupByID := make(map[uuidx.UUID]models.Group)
	for _, pair := range summary.Holdings {
		if pair.Holding.ValueKRW == nil {
			continue
		}
		totals[pair.Group.ID] = numeric.Wrap(totals[pair.Group.ID].Add(pair.Holding.ValueKRW.Decimal))
		groupByID[pair.Group.ID] = pair.Group
	}

	denominator := summary.TotalStockValue
	rows := make([]models.GroupSummaryRow, 0, len(totals))
	for id, total := range totals {
		group := groupByID[id]
		var actualPct numeric.Decimal
		if denominator.IsPositive() {
			actualPct = numeric.Wrap(total.Div(denominator.Decimal).Mul(hundred.Decimal))
		}
		targetPct, _ := numeric.FromString(fmt.Sprintf("%g", group.TargetPercentage))
		diffPct := numeric.Wrap(actualPct.Sub(targetPct.Decimal))
		diffVal := numeric.Wrap(total.Sub(denominator.Mul(targetPct.Decimal).Div(hundred.Decimal)))
		rows = append(rows, models.GroupSummaryRow{
			Group:     group,
			Total:     total,
			ActualPct: actualPct,
			TargetPct: targetPct,
			DiffPct:   diffPct,
			DiffVal:   diffVal,
		})
	}

	sort.Slice(rows, func(i, j int) bool {
		return rows[i].Total.GreaterThan(rows[j].Total.Decimal)
	})
	return rows
}

var hundred, _ = numeric.FromString("100")
